// corde/lib/src/expect/matches/role/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.ToSetRolePermission =
  exports.ToRenameRole =
  exports.ToSetRoleHoist =
  exports.ToSetRoleMentionable =
  exports.ToSetRolePosition =
  exports.ToDeleteRole =
  exports.ToSetRoleColor =
    void 0;

var toSetRoleColor_1 = require("./toSetRoleColor");

Object.defineProperty(exports, "ToSetRoleColor", {
  enumerable: true,
  get: function () {
    return toSetRoleColor_1.ToSetRoleColor;
  },
});

var toDeleteRole_1 = require("./toDeleteRole");

Object.defineProperty(exports, "ToDeleteRole", {
  enumerable: true,
  get: function () {
    return toDeleteRole_1.ToDeleteRole;
  },
});

var toSetRolePosition_1 = require("./toSetRolePosition");

Object.defineProperty(exports, "ToSetRolePosition", {
  enumerable: true,
  get: function () {
    return toSetRolePosition_1.ToSetRolePosition;
  },
});

var toSetRoleMentionable_1 = require("./toSetRoleMentionable");

Object.defineProperty(exports, "ToSetRoleMentionable", {
  enumerable: true,
  get: function () {
    return toSetRoleMentionable_1.ToSetRoleMentionable;
  },
});

var toSetRoleHoist_1 = require("./toSetRoleHoist");

Object.defineProperty(exports, "ToSetRoleHoist", {
  enumerable: true,
  get: function () {
    return toSetRoleHoist_1.ToSetRoleHoist;
  },
});

var toRenameRole_1 = require("./toRenameRole");

Object.defineProperty(exports, "ToRenameRole", {
  enumerable: true,
  get: function () {
    return toRenameRole_1.ToRenameRole;
  },
});

var toSetRolePermission_1 = require("./toSetRolePermission");

Object.defineProperty(exports, "ToSetRolePermission", {
  enumerable: true,
  get: function () {
    return toSetRolePermission_1.ToSetRolePermission;
  },
});
