// corde/lib/src/errors/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.QueueFunctionError =
  exports.TimeoutError =
  exports.TypeError =
  exports.PropertyError =
  exports.ConfigError =
  exports.FileError =
  exports.DiscordError =
  exports.CordeClientError =
    void 0;

var cordeClientError_1 = require("./cordeClientError");

Object.defineProperty(exports, "CordeClientError", {
  enumerable: true,
  get: function () {
    return cordeClientError_1.CordeClientError;
  },
});

var discordError_1 = require("./discordError");

Object.defineProperty(exports, "DiscordError", {
  enumerable: true,
  get: function () {
    return discordError_1.DiscordError;
  },
});

var fileError_1 = require("./fileError");

Object.defineProperty(exports, "FileError", {
  enumerable: true,
  get: function () {
    return fileError_1.FileError;
  },
});

var configError_1 = require("./configError");

Object.defineProperty(exports, "ConfigError", {
  enumerable: true,
  get: function () {
    return configError_1.ConfigError;
  },
});

var propertyError_1 = require("./propertyError");

Object.defineProperty(exports, "PropertyError", {
  enumerable: true,
  get: function () {
    return propertyError_1.PropertyError;
  },
});

var typeError_1 = require("./typeError");

Object.defineProperty(exports, "TypeError", {
  enumerable: true,
  get: function () {
    return typeError_1.TypeError;
  },
});

var timeoutError_1 = require("./timeoutError");

Object.defineProperty(exports, "TimeoutError", {
  enumerable: true,
  get: function () {
    return timeoutError_1.TimeoutError;
  },
});

var queueFunctionError_1 = require("./queueFunctionError");

Object.defineProperty(exports, "QueueFunctionError", {
  enumerable: true,
  get: function () {
    return queueFunctionError_1.QueueFunctionError;
  },
});
