// corde/lib/src/utils/colors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.resolveColor = exports.Colors = exports.ColorsHex = void 0;
var ColorsHex;

(function (ColorsHex) {
  ColorsHex[(ColorsHex["DEFAULT"] = 0)] = "DEFAULT";
  ColorsHex[(ColorsHex["WHITE"] = 16777215)] = "WHITE";
  ColorsHex[(ColorsHex["AQUA"] = 1752220)] = "AQUA";
  ColorsHex[(ColorsHex["GREEN"] = 3066993)] = "GREEN";
  ColorsHex[(ColorsHex["BLUE"] = 3447003)] = "BLUE";
  ColorsHex[(ColorsHex["YELLOW"] = 16776960)] = "YELLOW";
  ColorsHex[(ColorsHex["PURPLE"] = 10181046)] = "PURPLE";
  ColorsHex[(ColorsHex["LUMINOUS_VIVID_PINK"] = 15277667)] = "LUMINOUS_VIVID_PINK";
  ColorsHex[(ColorsHex["GOLD"] = 15844367)] = "GOLD";
  ColorsHex[(ColorsHex["ORANGE"] = 15105570)] = "ORANGE";
  ColorsHex[(ColorsHex["RED"] = 15158332)] = "RED";
  ColorsHex[(ColorsHex["GREY"] = 9807270)] = "GREY";
  ColorsHex[(ColorsHex["NAVY"] = 3426654)] = "NAVY";
  ColorsHex[(ColorsHex["DARK_AQUA"] = 1146986)] = "DARK_AQUA";
  ColorsHex[(ColorsHex["DARK_GREEN"] = 2067276)] = "DARK_GREEN";
  ColorsHex[(ColorsHex["DARK_BLUE"] = 2123412)] = "DARK_BLUE";
  ColorsHex[(ColorsHex["DARK_PURPLE"] = 7419530)] = "DARK_PURPLE";
  ColorsHex[(ColorsHex["DARK_VIVID_PINK"] = 11342935)] = "DARK_VIVID_PINK";
  ColorsHex[(ColorsHex["DARK_GOLD"] = 12745742)] = "DARK_GOLD";
  ColorsHex[(ColorsHex["DARK_ORANGE"] = 11027200)] = "DARK_ORANGE";
  ColorsHex[(ColorsHex["DARK_RED"] = 10038562)] = "DARK_RED";
  ColorsHex[(ColorsHex["DARK_GREY"] = 9936031)] = "DARK_GREY";
  ColorsHex[(ColorsHex["DARKER_GREY"] = 8359053)] = "DARKER_GREY";
  ColorsHex[(ColorsHex["LIGHT_GREY"] = 12370112)] = "LIGHT_GREY";
  ColorsHex[(ColorsHex["DARK_NAVY"] = 2899536)] = "DARK_NAVY";
  ColorsHex[(ColorsHex["BLURPLE"] = 7506394)] = "BLURPLE";
  ColorsHex[(ColorsHex["GREYPLE"] = 10070709)] = "GREYPLE";
  ColorsHex[(ColorsHex["DARK_BUT_NOT_BLACK"] = 2895667)] = "DARK_BUT_NOT_BLACK";
  ColorsHex[(ColorsHex["NOT_QUITE_BLACK"] = 2303786)] = "NOT_QUITE_BLACK";
})((ColorsHex = exports.ColorsHex || (exports.ColorsHex = {})));

var Colors;

(function (Colors) {
  Colors["DEFAULT"] = "#000000";
  Colors["WHITE"] = "#FFFFFF";
  Colors["AQUA"] = "#1ABC9C";
  Colors["GREEN"] = "#2ECC71";
  Colors["BLUE"] = "#3498DB";
  Colors["PURPLE"] = "#9B59B6";
  Colors["GOLD"] = "#E91E63";
  Colors["ORANGE"] = "#E67E22";
  Colors["RED"] = "#E74C3C";
  Colors["GREY"] = "#95A5A6";
  Colors["DARKER_GREY"] = "#7F8C8D";
  Colors["NAVY"] = "#34495E";
  Colors["DARK_AQUA"] = "#11806A";
  Colors["DARK_GREEN"] = "#1F8B4C";
  Colors["DARK_BLUE"] = "#206694";
  Colors["DARK_PURPLE"] = "#71368A";
  Colors["DARK_GOLD"] = "#C27C0E";
  Colors["DARK_ORANGE"] = "#A84300";
  Colors["DARK_RED"] = "#992D22";
  Colors["DARK_GREY"] = "#7F8C8D";
  Colors["LIGHT_GREY"] = "#BCC0C0";
  Colors["DARK_NAVY"] = "#2C3E50";
  Colors["LUMINOUS_VIVID_PINK"] = "#E91E63";
  Colors["DARK_VIVID_PINK"] = "#AD1457";
})((Colors = exports.Colors || (exports.Colors = {})));

function resolveColor(color) {
  if (color > 0 && color < 0xffffff) {
    return +color;
  }

  if (typeof color === "string" && color.includes("#")) {
    return parseInt(color.toString().replace("#", ""), 16);
  }

  if (Array.isArray(color)) {
    return (color[0] << 16) + (color[1] << 8) + color[2];
  }

  const value = ColorsHex[color];

  if (value) {
    return value;
  }

  return -1;
}

exports.resolveColor = resolveColor;
