// corde/lib/src/utils/permission.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.calcPermissionsValue = exports.permissionsArray = exports.Permission = void 0;
var Permission;

(function (Permission) {
  Permission[(Permission["CREATE_INSTANT_INVITE"] = 1)] = "CREATE_INSTANT_INVITE";
  Permission[(Permission["KICK_MEMBERS"] = 2)] = "KICK_MEMBERS";
  Permission[(Permission["BAN_MEMBERS"] = 4)] = "BAN_MEMBERS";
  Permission[(Permission["ADMINISTRATOR"] = 8)] = "ADMINISTRATOR";
  Permission[(Permission["MANAGE_CHANNELS"] = 16)] = "MANAGE_CHANNELS";
  Permission[(Permission["MANAGE_GUILD"] = 32)] = "MANAGE_GUILD";
  Permission[(Permission["ADD_REACTIONS"] = 64)] = "ADD_REACTIONS";
  Permission[(Permission["VIEW_AUDIT_LOG"] = 128)] = "VIEW_AUDIT_LOG";
  Permission[(Permission["VIEW_CHANNEL"] = 1024)] = "VIEW_CHANNEL";
  Permission[(Permission["SEND_MESSAGES"] = 2048)] = "SEND_MESSAGES";
  Permission[(Permission["SEND_TTS_MESSAGES"] = 4096)] = "SEND_TTS_MESSAGES";
  Permission[(Permission["MANAGE_MESSAGES"] = 8192)] = "MANAGE_MESSAGES";
  Permission[(Permission["EMBED_LINKS"] = 16384)] = "EMBED_LINKS";
  Permission[(Permission["ATTACH_FILES"] = 32768)] = "ATTACH_FILES";
  Permission[(Permission["READ_MESSAGE_HISTORY"] = 65536)] = "READ_MESSAGE_HISTORY";
  Permission[(Permission["MENTION_EVERYONE"] = 131072)] = "MENTION_EVERYONE";
  Permission[(Permission["USE_EXTERNAL_EMOJIS"] = 262144)] = "USE_EXTERNAL_EMOJIS";
  Permission[(Permission["CONNECT"] = 1048576)] = "CONNECT";
  Permission[(Permission["SPEAK"] = 2097152)] = "SPEAK";
  Permission[(Permission["MUTE_MEMBERS"] = 4194304)] = "MUTE_MEMBERS";
  Permission[(Permission["DEAFEN_MEMBERS"] = 8388608)] = "DEAFEN_MEMBERS";
  Permission[(Permission["MOVE_MEMBERS"] = 16777216)] = "MOVE_MEMBERS";
  Permission[(Permission["USE_VAD"] = 33554432)] = "USE_VAD";
  Permission[(Permission["PRIORITY_SPEAKER"] = 256)] = "PRIORITY_SPEAKER";
  Permission[(Permission["CHANGE_NICKNAME"] = 67108864)] = "CHANGE_NICKNAME";
  Permission[(Permission["MANAGE_NICKNAMES"] = 134217728)] = "MANAGE_NICKNAMES";
  Permission[(Permission["MANAGE_ROLES"] = 268435456)] = "MANAGE_ROLES";
  Permission[(Permission["MANAGE_WEBHOOKS"] = 536870912)] = "MANAGE_WEBHOOKS";
  Permission[(Permission["MANAGE_EMOJIS"] = 1073741824)] = "MANAGE_EMOJIS";
  Permission[(Permission["STREAM"] = 512)] = "STREAM";
  Permission[(Permission["VIEW_GUILD_INSIGHTS"] = 524288)] = "VIEW_GUILD_INSIGHTS";
})((Permission = exports.Permission || (exports.Permission = {})));

const permissionsArray = Object.keys(Permission);
exports.permissionsArray = permissionsArray;

function calcPermissionsValue(...permissions) {
  return permissions.reduce((p1, p2) => p1 | p2);
}

exports.calcPermissionsValue = calcPermissionsValue;
