// corde/lib/src/structures/cordeGuild.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.CordeGuild = void 0;

class CordeGuild {
  constructor(_guild) {
    this._guild = _guild;
  }

  get afkChannelID() {
    return this._guild.afkChannelID;
  }

  get afkTimeout() {
    return this._guild.afkTimeout;
  }

  get applicationID() {
    return this._guild.applicationID;
  }

  get approximateMemberCount() {
    return this._guild.approximateMemberCount;
  }

  get approximatePresenceCount() {
    return this._guild.approximatePresenceCount;
  }

  get isAvailable() {
    return this._guild.available;
  }

  get banner() {
    return this._guild.banner;
  }

  get isDeleted() {
    return this._guild.deleted;
  }

  get description() {
    return this._guild.description;
  }

  get discoverySplash() {
    return this._guild.discoverySplash;
  }

  get icon() {
    return this._guild.icon;
  }

  get id() {
    return this._guild.id;
  }

  get joinedAt() {
    return this._guild.joinedAt;
  }

  get isLarge() {
    return this._guild.large;
  }

  get maximumMembers() {
    return this._guild.maximumMembers;
  }

  get maximumPresences() {
    return this._guild.maximumPresences;
  }

  get membersCount() {
    return this._guild.memberCount;
  }

  get mfaLevel() {
    return this._guild.mfaLevel;
  }

  get name() {
    return this._guild.name;
  }

  get ownerId() {
    return this._guild.ownerID;
  }

  get isPartnered() {
    return this._guild.partnered;
  }

  get preferredLocale() {
    return this._guild.preferredLocale;
  }

  get premiumSubscriptionCount() {
    return this._guild.premiumSubscriptionCount;
  }

  get publicUpdatesChannelId() {
    return this._guild.publicUpdatesChannelID;
  }

  get region() {
    return this._guild.region;
  }

  get features() {
    return this._guild.features;
  }

  get isVerified() {
    return this._guild.verified;
  }

  get defaultMessageNotifications() {
    return this._guild.defaultMessageNotifications;
  }

  async delete() {
    await this._guild.delete();
  }

  async updateName(newName) {
    const guild = await this._guild.setName(newName);
    this._guild = guild;
    return this;
  }

  async updatePreferredLocale(newPreferredLocale) {
    const guild = await this._guild.setPreferredLocale(newPreferredLocale);
    this._guild = guild;
    return this;
  }

  async updateRegion(newRegion) {
    const guild = await this._guild.setRegion(newRegion);
    this._guild = guild;
    return this;
  }

  async updateBanner(newBanner) {
    const guild = await this._guild.setBanner(newBanner);
    this._guild = guild;
    return this;
  }

  async updateSplash(newSplash) {
    const guild = await this._guild.setSplash(newSplash);
    this._guild = guild;
    return this;
  }

  async updateVerificationLevel(verificationLevel) {
    const guild = await this._guild.setVerificationLevel(verificationLevel);
    this._guild = guild;
    return this;
  }

  async updateSystemChannelFlags(newSystemChannelFlags) {
    const guild = await this._guild.setSystemChannelFlags(newSystemChannelFlags);
    this._guild = guild;
    return this;
  }

  async updateWidget(newWidget) {
    this._guild = await this._guild.setWidget(newWidget);
    return this;
  }

  async updateSystemChannel(newChannel) {
    this._guild = await this._guild.setSystemChannel(newChannel);
    return this;
  }

  async updateRulesChannel(newChannel) {
    this._guild = await this._guild.setRulesChannel(newChannel);
    return this;
  }

  async updateRolePosition(...newRolePositions) {
    this._guild = await this._guild.setRolePositions(newRolePositions);
    return this;
  }

  async updatePublicUpdatesChannel(newPublicUpdatesChannel) {
    this._guild = await this._guild.setPublicUpdatesChannel(newPublicUpdatesChannel);
    return this;
  }

  async updateOwner(newOwner) {
    this._guild = await this._guild.setOwner(newOwner);
    return this;
  }

  async updateIcon(newIcon) {
    this._guild = await this._guild.setIcon(newIcon);
    return this;
  }

  async updateExplicitContentFilter(newExplicitContentFilter) {
    this._guild = await this._guild.setExplicitContentFilter(newExplicitContentFilter);
    return this;
  }

  async updateDiscoverySplash(newDiscoverySplash) {
    this._guild = await this._guild.setDiscoverySplash(newDiscoverySplash);
    return this;
  }

  async updateDefaultMessageNotifications(newDefaultMessageNotifications) {
    this._guild = await this._guild.setDefaultMessageNotifications(newDefaultMessageNotifications);
    return this;
  }

  async updateChannelPositions(...newChannelPositions) {
    this._guild = await this._guild.setChannelPositions(newChannelPositions);
    return this;
  }

  async updateAFKTimeout(newAfkTimeout) {
    this._guild = await this._guild.setAFKTimeout(newAfkTimeout);
    return this;
  }

  async updateAFKChannel(newAfkChannel) {
    this._guild = await this._guild.setAFKChannel(newAfkChannel);
    return this;
  }

  async update(newGuildData) {
    this._guild = await this._guild.edit(newGuildData);
    return this;
  }
}

exports.CordeGuild = CordeGuild;
