// corde/lib/src/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.TEST_FAIL_ICON =
  exports.TEST_PASSED_ICON =
  exports.TEST_RUNNING_ICON =
  exports.TAG_PASS =
  exports.TAG_FAIL =
  exports.TAG_PENDING =
  exports.TEXT_EMPTY =
  exports.TEXT_PENDING =
  exports.TEXT_FAIL =
  exports.TEXT_PASS =
  exports.DEFAULT_STACK_TRACE_LIMIT =
  exports.EXPECT_RECEIVED_TAB_SPACE =
  exports.MESSAGE_TAB_SPACE =
  exports.DEFAULT_TEST_TIMEOUT =
    void 0;

const tslib_1 = require("tslib");

const chalk_1 = tslib_1.__importDefault(require("chalk"));

exports.DEFAULT_TEST_TIMEOUT = 5000;
exports.MESSAGE_TAB_SPACE = "   ";
exports.EXPECT_RECEIVED_TAB_SPACE = "    ";
exports.DEFAULT_STACK_TRACE_LIMIT = 3;
exports.TEXT_PASS = chalk_1.default.green;
exports.TEXT_FAIL = chalk_1.default.red;
exports.TEXT_PENDING = chalk_1.default.yellow;
exports.TEXT_EMPTY = chalk_1.default.yellowBright;

const TAG_PENDING = (text = "RUNS") => chalk_1.default.bgYellow(chalk_1.default.black(` ${text} `));

exports.TAG_PENDING = TAG_PENDING;

const TAG_FAIL = (text = "FAIL") => {
  return chalk_1.default.bgRed(chalk_1.default.black(` ${text} `)) + " ";
};

exports.TAG_FAIL = TAG_FAIL;

const TAG_PASS = (text = "PASS") => {
  return chalk_1.default.bgGreen(chalk_1.default.black(` ${text} `)) + " ";
};

exports.TAG_PASS = TAG_PASS;
exports.TEST_RUNNING_ICON = "●";
exports.TEST_PASSED_ICON = exports.TEXT_PASS("✔ ");
exports.TEST_FAIL_ICON = exports.TEXT_FAIL("x");
