// corde/lib/src/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.Colors =
  exports.Permission =
  exports.beforeEach =
  exports.afterEach =
  exports.beforeStart =
  exports.afterAll =
  exports.createRole =
  exports.getRole =
  exports.sendMessage =
  exports.expect =
  exports.it =
  exports.describe =
  exports.test =
  exports.group =
    void 0;

var closures_1 = require("./closures");

Object.defineProperty(exports, "group", {
  enumerable: true,
  get: function () {
    return closures_1.group;
  },
});
Object.defineProperty(exports, "test", {
  enumerable: true,
  get: function () {
    return closures_1.test;
  },
});
Object.defineProperty(exports, "describe", {
  enumerable: true,
  get: function () {
    return closures_1.group;
  },
});
Object.defineProperty(exports, "it", {
  enumerable: true,
  get: function () {
    return closures_1.test;
  },
});

var expect_1 = require("./expect");

Object.defineProperty(exports, "expect", {
  enumerable: true,
  get: function () {
    return expect_1.expect;
  },
});

var api_utilities_1 = require("./api-utilities");

Object.defineProperty(exports, "sendMessage", {
  enumerable: true,
  get: function () {
    return api_utilities_1.sendMessage;
  },
});
Object.defineProperty(exports, "getRole", {
  enumerable: true,
  get: function () {
    return api_utilities_1.getRole;
  },
});
Object.defineProperty(exports, "createRole", {
  enumerable: true,
  get: function () {
    return api_utilities_1.createRole;
  },
});

var hooks_1 = require("./hooks");

Object.defineProperty(exports, "afterAll", {
  enumerable: true,
  get: function () {
    return hooks_1.afterAll;
  },
});
Object.defineProperty(exports, "beforeStart", {
  enumerable: true,
  get: function () {
    return hooks_1.beforeStart;
  },
});
Object.defineProperty(exports, "afterEach", {
  enumerable: true,
  get: function () {
    return hooks_1.afterEach;
  },
});
Object.defineProperty(exports, "beforeEach", {
  enumerable: true,
  get: function () {
    return hooks_1.beforeEach;
  },
});

var permission_1 = require("./utils/permission");

Object.defineProperty(exports, "Permission", {
  enumerable: true,
  get: function () {
    return permission_1.Permission;
  },
});

var colors_1 = require("./utils/colors");

Object.defineProperty(exports, "Colors", {
  enumerable: true,
  get: function () {
    return colors_1.Colors;
  },
});

const corde_1 = require("./corde");

exports.default = corde_1.corde;
