// corde/lib/src/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.SystemChannelFlag = exports.GuildFeatures = exports.VerificationLevel = void 0;
var VerificationLevel;

(function (VerificationLevel) {
  VerificationLevel["NONE"] = "NONE";
  VerificationLevel["LOW"] = "LOW";
  VerificationLevel["MEDIUM"] = "MEDIUM";
  VerificationLevel["HIGH"] = "HIGH";
  VerificationLevel["VERY_HIGH"] = "VERY_HIGH";
})((VerificationLevel = exports.VerificationLevel || (exports.VerificationLevel = {})));

var GuildFeatures;

(function (GuildFeatures) {
  GuildFeatures["INVITE_SPLASH"] = "INVITE_SPLASH";
  GuildFeatures["VIP_REGIONS"] = "VIP_REGIONS";
  GuildFeatures["VANITY_URL"] = "VANITY_URL";
  GuildFeatures["VERIFIED"] = "VERIFIED";
  GuildFeatures["PARTNERED"] = "PARTNERED";
  GuildFeatures["COMMUNITY"] = "COMMUNITY";
  GuildFeatures["COMMERCE"] = "COMMERCE";
  GuildFeatures["NEWS"] = "NEWS";
  GuildFeatures["DISCOVERABLE"] = "DISCOVERABLE";
  GuildFeatures["FEATURABLE"] = "FEATURABLE";
  GuildFeatures["ANIMATED_ICON"] = "ANIMATED_ICON";
  GuildFeatures["BANNER"] = "BANNER";
  GuildFeatures["WELCOME_SCREEN_ENABLED"] = "WELCOME_SCREEN_ENABLED";
  GuildFeatures["MEMBER_VERIFICATION_GATE_ENABLED"] = "MEMBER_VERIFICATION_GATE_ENABLED";
  GuildFeatures["PREVIEW_ENABLED"] = "PREVIEW_ENABLED";
})((GuildFeatures = exports.GuildFeatures || (exports.GuildFeatures = {})));

var SystemChannelFlag;

(function (SystemChannelFlag) {
  SystemChannelFlag["SUPPRESS_JOIN_NOTIFICATIONS"] = "SUPPRESS_JOIN_NOTIFICATIONS";
  SystemChannelFlag["SUPPRESS_PREMIUM_SUBSCRIPTIONS"] = "SUPPRESS_PREMIUM_SUBSCRIPTIONS";
})((SystemChannelFlag = exports.SystemChannelFlag || (exports.SystemChannelFlag = {})));
